core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2259,37 +2259,51 @@ private[spark] class DAGScheduler(
                     }
 
                     override def onShuffleMergeFailure(e: Throwable): Unit = {
+                      if (e.isInstanceOf[IOException]) {
+                        logInfo(s"Failed to connect external shuffle service " +
+                          s"${shuffleServiceLoc.hostPort}")
+                        blockManagerMaster.removeShufflePushMergerLocation(shuffleServiceLoc.host)
+                      }
                     }
                   })
             }
           }
         }, 0, TimeUnit.SECONDS)
       } else {
-        stage.shuffleDep.getMergerLocs.zipWithIndex.foreach {
-          case (shuffleServiceLoc, index) =>
-            // Sends async request to shuffle service to finalize shuffle merge on that host
-            // TODO: SPARK-35536: Cancel finalizeShuffleMerge if the stage is cancelled
-            // TODO: during shuffleMergeFinalizeWaitSec
-            shuffleClient.finalizeShuffleMerge(shuffleServiceLoc.host,
-              shuffleServiceLoc.port, shuffleId, shuffleMergeId,
-              new MergeFinalizerListener {
-                override def onShuffleMergeSuccess(statuses: MergeStatuses): Unit = {
-                  assert(shuffleId == statuses.shuffleId)
-                  eventProcessLoop.post(RegisterMergeStatuses(stage, MergeStatus.
-                    convertMergeStatusesToMergeStatusArr(statuses, shuffleServiceLoc)))
-                  results(index).set(true)
-                }
+        shuffleMergeFinalizeScheduler.schedule(new Runnable {
```

Comment:
Considering just 1 thread in the `shuffleMergeFinalizerScheduler`,  that thread will submit another task  (task to send RPCs) to pool here but then it will wait on line 2312 for the results. The thread will never get a chance to actually execute the other task (task which sends the RPCs).

Reply:
Thoughts on pushing finalization send into the threadpool instead ?

```
        stage.shuffleDep.getMergerLocs.zipWithIndex.foreach {
          case (shuffleServiceLoc, index) =>
            shuffleMergeFinalizeScheduler.schedule(new Runnable() {
              override def run(): Unit = {
                    // existing code within the "case"
               
              }
            }, 0, TimeUnit.SECONDS));
```

We should also bump up the default number of threads in this threadpool.

This will make sure that we wait utmost for `shuffleMergeResultsTimeoutSec` seconds for finalization to complete.

Thoughts ?

Reply:
Hi, @mridulm I don't think we should cancel all the `scheduledFutures` if there is a `TimeoutException`.  Merger ESS need this finalize RPC to finish its merge process. So  we should try our best to send all the RPCs even after `shuffleMergeResultsTimeoutSec`. Am I right ?

Reply:
When either all mergers report success or there is a timeout (whichever is earlier), we fire `ShuffleMergeFinalized` - at which point, the shuffle is marked finalized - and all subsequent merge statuses from ESS are ignored.

So sending `FinalizeShuffleMerge` to mergers does not impact the application once `ShuffleMergeFinalized` is done.

Reply:
For a long-running spark application, if no finalize partition RPC is received, ess will not close open data files, meta files, and index files until the application is complete. Too many opening files can be a potential risk. 

Or should we add a parameter to control whether to continue sending finalize RPC to the remaining nodes ?

Reply:
Agree, and that is why driver makes a best case effort to send the message.
If we are unable to do so in reasonable time due to network issues (n/w partition, etc), failures, and so on - there could be files which are not yet closed.
On other hand, keeping this pending state around in driver for extended periods of time will cause its own issues.

Note that in general, there will always be some dropped messages (for ex when failure handler is getting invoked) - so some variant of what you described can occur even if we make this specific case robust - what we have to evaluate is how much of an impact it has on shuffle service itself in context of  whether it will negatively impact NM stability.

In our environment, this has been in prod for a while now - and we have not seen this specific issue. Having said that, for a different application and network characteristics, it could possibly be an issue : more information will help understand it.

Reply:
Another idea, could we stop sending finalize RPCs after `shuffleMergeResultsTimeoutSec`. And start a thread in ESS that periodically checks and closes the opened files which the shuffle has already finalized?
Do that in another PR?
What do you think? @mridulm @otterc

Reply:
ESS does not connect back to the executors or driver - connections/requests are  initiated by executors and driver to ESS.